core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailRepositoryConnection.java

```diff
@@ -262,7 +269,22 @@ public SailBooleanQuery prepareBooleanQuery(QueryLanguage ql, String queryString
 	public Update prepareUpdate(QueryLanguage ql, String update, String baseURI)
 			throws RepositoryException, MalformedQueryException {
 		ParsedUpdate parsedUpdate = QueryParserUtil.parseUpdate(ql, update, baseURI);
+		List<UpdateExpr> updateExprs = parsedUpdate.getUpdateExprs();
+		SPARQLUpdateDataBlockParser parser = new SPARQLUpdateDataBlockParser(this.getValueFactory());
+		for (UpdateExpr expr : updateExprs) {
+			String datablock = "";
+			if (expr instanceof InsertData) {
+				datablock = ((InsertData) expr).getDataBlock();
+			} else if (expr instanceof DeleteData) {
+				datablock = ((DeleteData) expr).getDataBlock();
+			}
```

Comment:
I was thinking that perhaps we should move all of this into the SPARQLParser itself. It seems the more logical place to handle it.Then again doing it here gives us a bit more flexibility. Fine either way.

When actually executing the insert/delete, the parser is reconfigured for specific types of errors. For INSERTs, the following config settings are done:

   		parser.setLineNumberOffset(insertDataExpr.getLineNumberOffset());
		parser.getParserConfig().addNonFatalError(BasicParserSettings.VERIFY_DATATYPE_VALUES);
		parser.getParserConfig().addNonFatalError(BasicParserSettings.FAIL_ON_UNKNOWN_DATATYPES);
		parser.getParserConfig().set(BasicParserSettings.SKOLEMIZE_ORIGIN, null);

and for DELETE, the following:

		parser.setLineNumberOffset(deleteDataExpr.getLineNumberOffset());
		parser.setAllowBlankNodes(false); // no blank nodes allowed in DELETE DATA.

I don't think we need the 'non-fatal error' settings (after all we _want_ a parse exception when an error occurs), but the other settings we should probably duplicate here. The line number offset will hopefully correct the odd line numbers you're currently seeing in the tests.

Reply:
Hey @jeenbroekstra, thanks for the valuable feedback. We are facing the following error when adding SPARQLUpdateDataBlockParser in the SPARQLParser.java file - 
"Adding dependency on module 'rdf4j-repository-sail' will introduce **circular dependency** between modules 'rdf4j-queryalgebra-evaluation' and 'rdf4j-queryparser-sparql'" .

Let us know how to proceed. Meanwhile, we are trying to figure out some other way to parse datablocks.

Reply:
@ishaanbassi Ah good point, I hadn't thought of that. This will require a bit of refactoring. I suggest you move the SPARQLUpdateDatablockParser class to package `org.eclipse.rdf4j.query.parser.sparql`, in the `rdf4j-queryparser-sparql` module. That should solve the circular dependency issue. You will also need to add a new dependency for the trig parser to the pom.xml in the `rdf4j-queryparser-sparql` module:

		<dependency>
			<groupId>${project.groupId}</groupId>
			<artifactId>rdf4j-rio-trig</artifactId>
			<version>${project.version}</version>
		</dependency>

To preserve backward compatibility, leave an empty class with the old name in the old location, marked `@deprecated`.  

You know, I was actually doing these change while writing this, to make sure I wasn't making a mistake. I might as well push to your branch, so you can see what I mean.